casper/src/main/scala/coop/rchain/casper/engine/BlockApproverProtocol.scala

```diff
@@ -178,11 +177,31 @@ object BlockApproverProtocol {
         _ <- (blockDeploys.size == genesisBlessedContracts.size)
               .either(())
               .or("Mismatch between number of candidate deploys and expected number of deploys.")
+
+        // check if expected blessed contracts and ones reported by candidate are equal
+        (leftover, _) = genesisBlessedContracts.toList.foldLeft(
+          (genesisBlessedContracts, false)
+        ) {
+          case ((leftover, inconsistencyFound), blessedDeploy) =>
+            if (inconsistencyFound) (leftover, inconsistencyFound)
+            else {
+              blockDeploys.find(_.deploy.data.term == blessedDeploy.data.term) match {
+                case Some(_) =>
+                  (leftover - blessedDeploy, false)
+                case None =>
+                  (leftover, true)
+              }
+            }
+        }
+        _ <- (leftover.isEmpty)
+              .either(())
+              .or("Genesis candidate deploys do not match expected blessed contracts.")
       } yield (blockDeploys, block.body.state)
 
     (for {
       result                    <- EitherT(validate.pure[F])
       (blockDeploys, postState) = result
+      // check if candidate blessed contracts match candidate postStateHash
       stateHash <- EitherT(
                     runtimeManager
                       .replayComputeState(runtimeManager.emptyStateHash)(
```

Comment:
Why we didn't see different hash when genesis validator replayed a block with different blessed contracts?

Reply:
Because `blockDeploys` here are deploys from candidate block itself. So there were no check at all that local config match candidate block received. The only check was that block received is replayable, even if blessed contracts inside candidate do not match our expectation.
https://github.com/rchain/rchain/blob/a89a137ef094f0873c1370f0aa52c6c06c9a1995/casper/src/main/scala/coop/rchain/casper/engine/BlockApproverProtocol.scala#L199-L212

Reply:
We cannot create genesis ourselves to check if hash matches candidate - due to nondeterminism hash will be different. Se have are limited to replaying processed deploys with event logs from candidate.  But then we need to be sure that code being replayed is actually what we expect. This is added in PR.

Reply:
We need to have cryptographic proof so any other comparison is a convenience to report more detailed errors.

Let's acknowledge what genesis validator needs to be able to replay the block in the same way as ceremony master.
It needs the same deploys and system deploys - we know we have it because we should run the same code. As you mentioned, this is not enough to replay due to nondeterminism in Rholang.
But we have event logs from ceremony master block as the second part of input which is then enough to successfully replay a block and get the final hash for crypto proof.

Reply:
Why not make as its now? To compare terms that node is expect with terms that relayed?

We can use event logs from processed deploys in candidate to recreate processed deploy from local blessed contracts - but that is effectively the same operation? Except it is more relaxed in a way that terms should not be equal up to string, it should be just the same rholang code, possibly formatted differently.

Reply:
Actually that would be simpler to just substitute term, let me try

Reply:
If you are sure about that, can I be a ceremony master? :smiley: What if I execute some additional code not as part of deploy s and system deploys?

Reply:
You won't be able to execute it because I will check if terms in list of processed deploys from candidate you've sent me match terms that I expect to be in genesis. And number of expected terms equal number of processed deploys. So your processed deploys have to contain exactly the same blessed contracts that I expect to see. And then I verify that terms you've put in processed deploys create state with hash that you've claimed. So I can be sure state is valid and match my view on genesis.

Reply:
And if I add some additional code in one of the system deploys?

Reply:
System deploys in genesis are empty for genesis
https://github.com/rchain/rchain/blob/a89a137ef094f0873c1370f0aa52c6c06c9a1995/casper/src/main/scala/coop/rchain/casper/engine/BlockApproverProtocol.scala#L205-L213

Reply:
Per our discussion on standup. @nzpr convinced me, and Greg confirmed, that only deploys should be initial inputs to genesis (which is great) so equality check is all that is needed to validate before replay.